File: Easy Problems/Leetcode 20.cpp
class Solution {
public:
    bool isValid(string s) {
        if (s.size() % 2 != 0) return false;
        
        stack<char> st;
        for (char ch : s) {
            
            if (ch == '(' || ch == '[' || ch == '{') {
                st.push(ch);
            } 
            
            else { 
                if (st.empty()) {
                    return false;
                }
                
                if (ch == ')' && st.top() == '(') {
                    st.pop();
                } else if (ch == ']' && st.top() == '[') {
                    st.pop();
                } else if (ch == '}' && st.top() == '{') {
                    st.pop();
                } else {
                    return false;
                }
            }
        }
        
        return st.empty();
    }
};

File: Easy Problems/Leetcode 20.py
class Solution:
    def isValid(self, s: str) -> bool:
        st = []
        for ch in s:
            if ch in '{[(':
                st.append(ch)
            elif len(st) == 0 or (ch == ')' and st[-1] != '(') or (ch == '}' and st[-1] != '{') or (ch == ']' and st[-1] != '['):
                return False
            else:
                st.pop()
        return not len(st)